sdks/java/harness/src/main/java/org/apache/beam/fn/harness/Caches.java

```diff
@@ -0,0 +1,285 @@
+      }
+
+      @Override
+      public void remove(Object key) {
+        cache.remove(key);
+      }
+
+      @Override
+      public Set<Object> keys() {
+        return cache.keys();
+      }
+    };
+  }
+
+  /**
+   * Additional operations necessary for a top level cache to support sub-caching.
+   *
+   * <p>Visibility is restricted to prevent usage of these methods outside of this class.
+   */
+  private interface SubCacheable<K, V> extends Cache<K, V> {
+    Set<K> keys();
+  }
+
+  /**
+   * A view of a cache that operates on keys with a specified key prefix.
+   *
+   * <p>All lookups, insertions, and removals into the parent {@link Cache} will be prefixed by the
+   * specified prefixes.
+   *
+   * <p>Operations which operate over the entire caches contents such as {@link Cache#clear} only
+   * operate over keys with the specified prefixes.
+   */
+  private static class SubCache<K, V> implements Cache<K, V> {
+    private final SubCacheable<CompositeKey, Object> cache;
+    private final CompositeKey keyPrefix;
+
+    SubCache(SubCacheable<?, ?> cache, CompositeKey keyPrefix) {
+      this.cache = (SubCacheable<CompositeKey, Object>) cache;
+      this.keyPrefix = keyPrefix;
+    }
+
+    @Override
+    public V peek(K key) {
+      return (V) cache.peek(keyPrefix.subKey(key));
+    }
+
+    @Override
+    public V computeIfAbsent(K key, Function<K, V> loadingFunction) {
+      return (V)
+          cache.computeIfAbsent(
+              keyPrefix.subKey(key),
+              new Function<CompositeKey, Object>() {
+
+                @Override
+                public Object apply(CompositeKey o) {
+                  return loadingFunction.apply((K) o.keyParts[o.keyParts.length - 1]);
+                }
+              });
+    }
+
+    @Override
+    public void put(K key, V value) {
+      cache.put(keyPrefix.subKey(key), value);
+    }
+
+    @Override
+    public void clear() {
+      for (Object key : Sets.filter(cache.keys(), (Object o) -> keyPrefix.isProperPrefixOf(o))) {
+        cache.remove((CompositeKey) key);
+      }
+    }
+
+    @Override
+    public void remove(K key) {
+      cache.remove(keyPrefix.subKey(key));
+    }
+  }
+
+  /** A tuple of key parts used to represent a key within a cache. */
+  @VisibleForTesting
+  static class CompositeKey {
+    public static final CompositeKey ROOT = new CompositeKey(new Object[0]);
+    Object[] keyParts;
+
+    private CompositeKey(Object[] keyParts) {
+      this.keyParts = keyParts;
+    }
+
+    CompositeKey subKey(Object suffix, Object... additionalSuffixes) {
+      Object[] subKey = new Object[keyParts.length + 1 + additionalSuffixes.length];
+      System.arraycopy(keyParts, 0, subKey, 0, keyParts.length);
+      subKey[keyParts.length] = suffix;
+      System.arraycopy(
+          additionalSuffixes, 0, subKey, keyParts.length + 1, additionalSuffixes.length);
+      return new CompositeKey(subKey);
+    }
+
+    boolean isProperPrefixOf(Object possiblySuffix) {
```

Comment:
Nit: I don't think "suffix" is the right word for the argument here (if A is a prefix of B, then `B = A + suffix`) but I'm unsure of what the right word is. See https://english.stackexchange.com/questions/421424/what-is-the-opposite-of-a-prefix